letsencrypt/client/dns_authenticator.py

```diff
@@ -0,0 +1,261 @@
+            % (zone)),
+        8: ('Records that should exist do not exist (_acme-challenge.%s)'
+            % (zone)),
+        9: ('Server is not authorized or is using bad TSIG key to '
+            'make updates to zone "%s"' % (zone)),
+        10: ('Zone "%s" does not exist' % (zone)),
+        16: ('Server is using bad TSIG key to '
+                    'make updates to zone "%s"' % (zone)),
+        17: ('Server is using bad TSIG key to '
+                    'make updates to zone "%s"' % (zone))
+        }
+
+    try:
+        response = dns.query.tcp(
+            dns_request,
+            server, port=port,
+            source_port=constants.DNS_CHALLENGE_SOURCE_PORT,
+            timeout=constants.DNS_CHALLENGE_TIMEOUT
+        )
+
+        if response.rcode() == 0:
+            return True
+        else:
+            raise errors.LetsEncryptDNSAuthError(
+                rcode_errors.get("DNS Error: %s" % (response.rcode())))
+
+    except (
+        dns.resolver.NoAnswer,
+        dns.query.UnexpectedSource,
+        dns.query.BadResponse,
+        OSError
+    ) as err: # TimeoutError doesn't exist in 2.7 afaik
+        # elif isinstance(err, TimeoutError):
+        #     dns_error = "DNS Error: DNS request timed out!"
+        if isinstance(err, dns.resolver.NoAnswer):
+            dns_error = "DNS Error: Did not recieve a response to DNS request!"
+        elif isinstance(err, dns.query.UnexpectedSource):
+            dns_error = ("DNS Error: Recieved response to DNS request from "
+                         "unexpected source!")
+        elif isinstance(err, dns.query.BadResponse):
+            dns_error = ("DNS Error: Recieved malformed response to DNS "
+                         "request!")
+        elif isinstance(err, OSError):
+            dns_error = ("DNS Error: I forgot what an OSError means in this"
+                         " context...")
+        raise errors.LetsEncryptDNSAuthError(dns_error)
+
+class DNSAuthenticator(object):
+    """DNS Authenticator
+
+    This authenticator interacts with an already configured DNS
+    server in order to provision subdomains in response to DNS
+    challenges from the certificate authority.
+
+    """
+    zope.interface.implements(interfaces.IAuthenticator)
+
+    def __init__(self, config):
+        # super(DNSAuthenticator, self).__init__(config)
+        self.dns_server = config.dns_server
+        self.dns_server_port = config.dns_server_port
+        self.dns_tsig_keys = config.dns_tsig_keys
+
+    @staticmethod
+    def get_chall_pref(unused_domain):
+        """Get challenge preferences.
+
+        :returns: A list containing only 'challenges.DNS'.
+
+        """
+        return [challenges.DNS]
+
+    def perform(self, achalls):
+        """Perform the challenges.
+
+        """
+        if not achalls or not isinstance(achalls, list):
+            raise ValueError(".perform() was called without challenge list")
```

Comment:
while I can see that you based your work on `standalone_authenticator`, this check is not necessary, as `IAuthenticator.perform` is guaranteed to receive non-empty list of expected challenges only (http://letsencrypt.readthedocs.org/en/latest/api/client/interfaces.html#letsencrypt.client.interfaces.IAuthenticator.perform)